### ISSUE TITLE
Random VS/Mac compile failures for sharing violation

### ISSUE THREAD

**Issue by FloridaMatt** [AvaloniaUI/Avalonia#5294]
MacOS Big Sur 11.1
Visual Studio for Mac 8.8.5 (build 18)
Avalonia 0.9.12

About a quarter of the time after correcting a syntax error that caused a build failure, the next build attempt fails with the following message:

/Users/t/.nuget/packages/avalonia/0.9.12/build/AvaloniaBuildTasks.targets(5,5): Error MSB4018: The "CompileAvaloniaXamlTask" task failed unexpectedly.
System.IO.IOException: Sharing violation on path /Users/t/Projects/PhotoViewer/obj/Debug/netcoreapp5.0/osx-x64/Avalonia/original.pdb
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00259] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:274 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.IO/FileStream.cs:106 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.FileSystem.CopyFile (System.String sourceFullPath, System.String destFullPath, System.Boolean overwrite) [0x00035] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:55 
  at System.IO.File.Copy (System.String sourceFileName, System.String destFileName, System.Boolean overwrite) [0x00056] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.IO.FileSystem/src/System/IO/File.cs:74 
  at Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute () [0x00074] in D:\a\1\s\src\Avalonia.Build.Tasks\CompileAvaloniaXamlTask.cs:29 
  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x00023] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs:577 
  at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask (Microsoft.Build.BackEnd.ITaskExecutionHost taskExecutionHost, Microsoft.Build.BackEnd.Logging.TaskLoggingContext taskLoggingContext, Microsoft.Build.BackEnd.TaskHost taskHost, Microsoft.Build.BackEnd.ItemBucket bucket, Microsoft.Build.BackEnd.TaskExecutionMode howToExecuteTask) [0x002b9] in /Users/builder/jenkins/workspace/build-package-osx-mono-pullrequest/pr/external/bockbuild/builds/msbuild-15/src/Build/BackEnd/Components/RequestBuilder/TaskBuilder.cs:809  (MSB4018) (PhotoViewer)

The only workaround I know of is to close and reopen the project, after which it compiles successfully (if there are no other errors),

**Comment by trympet** [comment]
I was able to reproduce the issue when compiling with the `/maxCpuCount` flag set

**Comment by CBesser** [comment]
I've seen this too, lsof's reporting `mono-sgen` holds the file lock

**Comment by mrj001** [comment]
I'm also seeing this happen.
MacOS: 10.15.7
Visual Studio for Mac: 8.10.11 (build 8)
Dotnet version: 5.0.402

It does not happen if I build with either "dotnet build" at the command line or if I build with Visual Studio Code. (I think these two are equivalent.)

So there's something different about the way VS/Mac builds vs. VSCode/Command Line.

**Comment by Tlaster** [comment]
I'm also seeing this on Linux
5.16.14-1-MANJARO
dotnet 6.0.201
Just create a new Avalonia cross-platform project by `dotnet new avalonia.xplat` and run `dotnet run` in the Android project will get this error, but `dotnet build` seems to be fine

**Comment by newburner** [comment]
still getting this error
on arch linux 5.19.4 
rider /vs code /cli
dotnet 6.0.400

im getting this error even on a completely new project
anyone found a soultion?

**Comment by asvol** [comment]
I had a similar error when debug for Android. 
Host: Win10
Avalonia: 11.0.0-rc1.1
ReactiveUI.Fody: 19.2.1
Rider + .NET 7.0

I was able to fix this error by disabling parallel build (maxCpuCount=1). 

<img width="942" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/1770739/865d5f17-e73b-4e38-8e4e-264ca08317b9">

<img width="737" alt="image" src="https://github.com/AvaloniaUI/Avalonia/assets/1770739/13f9558d-eead-4043-8e95-ffb1162ba1e4">

I think it happen when using ReactiveUI.Fody packet:  it generate for PropertyChanges code for marked [Reactive] attributes properties when build is runing.

**Comment by timunie** [comment]
Now that we have the CommunityToolkit.MVVM package, the `[Reactive]` attribute can be replaced by `[ObservableProperty]` which is a source generator and thus seems to perform better in most cases.

**Comment by Snegovikufa** [comment]
If you use library classes and not using any UI/XAML, the problem could be solved by disabling the compilation of XAML:
`<EnableAvaloniaXamlCompilation>false</EnableAvaloniaXamlCompilation>`

**Comment by DerRiedi** [comment]
Disabling "Use ReSharper Build" does fix it on Ubuntu 22.04 / JB Rider / .NET 7.0

**Comment by buzzware** [comment]
I'm getting this on MacOS / Rider / .Net 7 / Android emulator

Simply disabling "Use ReSharper Build" and debug running again did not fix it.

**Alternating between Build "Use up to [ 14 ] processes in parallel" VS [ 1 ] does show that 14 triggers the problem**

```
  AvaloniaBuildTasks.targets(120, 5): [MSB4018] The "CompileAvaloniaXamlTask" task failed unexpectedly.
System.IO.IOException: The process cannot access the file '/Users/gary/repos/civmec/CivtracSupervisor/AvaloniaInsideShell/src/AvaloniaInside.Shell/obj/Debug/net7.0/Avalonia/original.pdb' because it is being used by another process.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Init(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Int64& fileLength, UnixFileMode& filePermissions)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Func`4 createOpenException)
   at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() in /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:line 35
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)

```

**Comment by lsoft** [comment]
Hi!

We have faced with this issue too. To be honest every 3rd compilation pipeline fails on our compilation servers so the working process has stopped with endless reruns. We are in fire with this bug...

```
C:\Users\<user>\.nuget\packages\avalonia\11.0.4\buildTransitive\AvaloniaBuildTasks.targets(120,5): error MSB4018: непредвиденная ошибка при выполнении задачи "CompileAvaloniaXamlTask".
System.IO.IOException: Процесс не может получить доступ к файлу "obj\.sual\Debug\net7.0\Avalonia\original.pdb", так как этот файл используется другим процессом.
   в System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   в System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   в Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() в /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:строка 36
   в Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   в Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [<here is the path to failed csproj>]
```

**Comment by lsoft** [comment]
We can easily reproduce it with the following powershell script

```
Set-Location 'folder of sln file';

$Iteration = 0;
while(1)
{
    [System.Console]::WriteLine('Iteration: ' + $Iteration + ', Now: ' + [System.DateTime]::Now);

    $MsBuildProcess = Start-Process -WindowStyle Hidden -Wait -PassThru -FilePath 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\MSBuild.exe' -ArgumentList ( ' your.sln -bl:your.binlog' );
    if($MsBuildProcess.ExitCode -ne 0)
    {
        [System.Console]::WriteLine('msbuild fail!');
        exit $MsBuildProcess.ExitCode;
    }

    $Iteration += 1;
}
```

wait for this script will finish with `msbuild fail` and take a look inside `your.binlog` (by using an appropriate utility).

**Comment by lsoft** [comment]
@kekekeks @maxkatz6  or other maintainers: here is almost 100% repro , the root cause is the multilply `root` projects, all of them is referencing (via others csprojes) to `BBBBBB` project, and `BBBBBB` project fails with 

```
C:\Users\<username>\.nuget\packages\avalonia\11.0.4\buildTransitive\AvaloniaBuildTasks.targets(120,5): error MSB4018: непредвиденная ошибка при выполнении задачи "CompileAvaloniaXamlTask".
System.IO.IOException: Процесс не может получить доступ к файлу "obj\Debug\net7.0\Avalonia\original.pdb", так как этот файл используется другим процессом.
   в System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   в System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)
   в Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() в /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:строка 36
   в Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   в Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
```

[master_A2Z.zip](https://github.com/AvaloniaUI/Avalonia/files/13197489/master_A2Z.zip)


Interestingly, compilation inside VS is completely fine, so one needs to compile from command line to see the bug. Here is  repro powershell script:

```
Set-Location 'path to master_A2Z folder';

$RestoreProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'dotnet' -ArgumentList ( ' restore A2Z.sln ' );
Wait-Process -InputObject $RestoreProcess;

$Iteration = 0;
while(1)
{
    [System.Console]::WriteLine('Iteration: ' + $Iteration + ', Now: ' + [System.DateTime]::Now);

    # probably you will need to change this msbuild path
    $MsBuildProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Current\Bin\amd64\MSBuild.exe' -ArgumentList ( ' A2Z.sln /v:quiet /bl:mk.binlog /maxcpucount:1 /p:BuildInParallel=false' );
    Wait-Process -InputObject $MsBuildProcess;
    if($MsBuildProcess.ExitCode -ne 0)
    {
        [System.Console]::WriteLine('msbuild fail!');
        exit $MsBuildProcess.ExitCode;
    }

    $Iteration += 1;
}
```

running this, I see the following:

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/29/2023 12:33:42
msbuild fail!
```

or rarely this:

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/29/2023 12:26:08
Iteration: 1, Now: 10/29/2023 12:32:07
msbuild fail!
```


looking at `MSBuild Structured Log Viewer` I see this error.

here is Process Monitor log

```
13:01:18,3088972	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Read/Write, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3092621	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3096137	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: None, AllocationSize: 10 676
13:01:18,3099288	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3102536	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3105866	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Delete, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3109093	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3113714	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Data/List Directory, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:18,3118344	MSBuild.exe	142976	IRP_MJ_CREATE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SHARING VIOLATION	Desired Access: Generic Write, Read Attributes, Write DAC, Disposition: OverwriteIf, Options: Sequential Access, Synchronous IO Non-Alert, Non-Directory File, Attributes: A, ShareMode: Read, Write, AllocationSize: 10 676
13:01:23,1389123	MSBuild.exe	142976	IRP_MJ_CLEANUP	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
13:01:23,1389723	MSBuild.exe	142976	IRP_MJ_CLOSE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
13:01:23,4386530	System	4	FASTIO_ACQUIRE_FOR_SECTION_SYNCHRONIZATION	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	SyncType: SyncTypeOther
13:01:23,4387064	System	4	IRP_MJ_CLOSE	C:\<here is the path>\master_A2Z\src\A2Z\BBBBBB\obj\Debug\net7.0\Avalonia\original.pdb	SUCCESS	
```

if I read this correctly, MSBuild conflicts with itself (PTAL on `/maxcpucount:1 /p:BuildInParallel=false` in powershell script).


I hope any maintainer will take a look here, because this is very difficult problem for those who faced it. We are going to perform autorerun if  building fails and `original.pdb` contains in build log, but our compilation cluster will degrade severely because of this.

**Please let me know if you will need any info\support.**

**Comment by maxkatz6** [comment]
Please let us know if this issue is solved in the latest master (or future releases). This issue isn't reproducible on our machines, and we only can guess that it could be due to different GC releasing file handles at different times. Recent PR should make releasing file handles a solved issue.

**Comment by lsoft** [comment]
@maxkatz6 I switch my repro sln to the version `11.1.999-cibuild0041387-beta` and see the following log

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/31/2023 08:53:59
Iteration: 1, Now: 10/31/2023 08:54:37
Iteration: 2, Now: 10/31/2023 08:55:00
Iteration: 3, Now: 10/31/2023 08:55:27
Iteration: 4, Now: 10/31/2023 08:55:53
Iteration: 5, Now: 10/31/2023 08:56:15
Iteration: 6, Now: 10/31/2023 08:56:38
Iteration: 7, Now: 10/31/2023 08:57:01
Iteration: 8, Now: 10/31/2023 08:57:29
Iteration: 9, Now: 10/31/2023 08:57:53
Iteration: 10, Now: 10/31/2023 08:58:19
Iteration: 11, Now: 10/31/2023 08:58:43
Iteration: 12, Now: 10/31/2023 08:59:07
Iteration: 13, Now: 10/31/2023 08:59:32
Iteration: 14, Now: 10/31/2023 08:59:56
Iteration: 15, Now: 10/31/2023 09:00:16
Iteration: 16, Now: 10/31/2023 09:00:37
Iteration: 17, Now: 10/31/2023 09:00:59
Iteration: 18, Now: 10/31/2023 09:01:26
Iteration: 19, Now: 10/31/2023 09:01:49
```

then I swith it to `11.0.5` **back** and see this:

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 10/31/2023 09:03:45
msbuild fail!
```

so I guess **issue is resolved**. I will let you know if anything will go wrong again.

@kekekeks @maxkatz6 thank you guys for rapid reaction. we saved from a lot of pain... is there any plan to release 11.0.6 or something like it? this will allow us to switch away from nightly feed...

ps. perhaps you want to close a related issue https://github.com/AvaloniaUI/Avalonia/issues/12670

**Comment by lsoft** [comment]
@maxkatz6 @kekekeks the bug has returned.

![image](https://github.com/AvaloniaUI/Avalonia/assets/5988558/89550216-0473-4a1d-97a4-560efe3816af)

I decompiled (dotpeek) the file `C:\Users\<user>\.nuget\packages\avalonia\11.1.999-cibuild0041411-beta\tools\netstandard2.0` and recent fix is in place:

![image](https://github.com/AvaloniaUI/Avalonia/assets/5988558/a22aa28d-800b-4733-be53-5d0b04d62898)

Please reopen this issue. We need your help again :(

could I do anything to help the investigation?

**Comment by lsoft** [comment]
I want to add one more point. Something has changed after @kekekeks `using` fix.

My message above is from repo I shared earlier. But in our commercial system I see sometimes only

```
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Current\Bin\amd64\Microsoft.Common.CurrentVersion.targets(4702,5): error MSB3030: не удалось скопировать файл "obj\Release\net7.0\my.dll" - файл не найден. [E:\glr\builds\kryMeSns\0\...\Mobile\Avalonia\....\our avalonia using.csproj]
```

with no `original.pdb` at all in msbuild binlog.

Looks like there is additional issue.

**Comment by lsoft** [comment]
@maxkatz6 @kekekeks looks like I found what is wrong now.

I get [master_A2Z.zip](https://github.com/AvaloniaUI/Avalonia/files/13197489/master_A2Z.zip) repro, run `dotnet restore` and run the following script:

```
Set-Location 'folder path for master_A2Z';

$Iteration = 0;
while(1)
{
    [System.Console]::WriteLine('Iteration: ' + $Iteration + ', Now: ' + [System.DateTime]::Now);

    #$MsBuildProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'dotnet' -ArgumentList ( ' build -m:1 /nr:false -clp:"ErrorsOnly;Summary" /p:Platform="Any CPU";Configuration="Release";Optimize="true";DebugSymbols="true";DebugType="full";BuildInParallel=false /bl:avalonia.binlog' );
    $MsBuildProcess = Start-Process -WindowStyle Hidden -PassThru -FilePath 'dotnet' -ArgumentList ( ' build -m /nr:false -clp:"ErrorsOnly;Summary" /p:Platform="Any CPU";Configuration="Release";Optimize="true";DebugSymbols="true";DebugType="full" /bl:avalonia.binlog' );
    Wait-Process -InputObject $MsBuildProcess;
    if($MsBuildProcess.ExitCode -ne 0)
    {
        [System.Console]::WriteLine('msbuild fail!');
        exit $MsBuildProcess.ExitCode;
    }

    $Iteration += 1;
}
```

and I see the error. then I comment `$MsBuildProcess = Start-Process` line, and UNcomment `#$MsBuildProcess = Start-Process` line, and rerun the script. Everything was OK for 20 iterations. Then I stopped the script and return that `wrong` line back and see the error again:

![image](https://github.com/AvaloniaUI/Avalonia/assets/5988558/193e6ff7-baa8-40fb-8384-5130e74b1466)

the difference between lines is `-m` <-> `-m:1 BuildInParallel=false `!

So, things look like that @kekekeks's PR fixed the bug in sequential compilation (which totally makes sense!), but something wrong is still here for parallel compilation!

Hope this helps you to resolve. At least, we have viable workaround now.

We are keep eye on this bug. Sorry for bothering :)

**Comment by buzzware** [comment]
> @kekekeks @maxkatz6 thank you guys for rapid reaction. we saved from a lot of pain... is there any plan to release 11.0.6 or something like it? this will allow us to switch away from nightly feed...

Yes indeed - this issue is making life hard, and its been over 4 weeks since 11.0.5.

**Comment by buzzware** [comment]
What is the best version of Avalonia right now - 11.0.5 or which nightly?

**Comment by timunie** [comment]
@buzzware as the issue is still open I'm not sure it is fixed in nightly, If it is fixed there, I'd go with nightly tbh. 

Moreover we have the label help-wanted. So if anyone has an idea and the needed machine to test, please file a PR.

**Comment by lsoft** [comment]
@buzzware 

> 11.0.5 or which nightly?

we are using some nightly revision. everything looks good, but we are waiting for 11.0.6 to have additional safety....

@timunie 

>  if anyone has an idea and the needed machine to test, please file a PR

let me ask: I had shared the repro above. did Avalonia team check that repro? the repro has failed for me with almost 100% probability. Let me know if you have any troubles with that repro, because I'm highly motivated to see this issue resolved.

**Comment by lsoft** [comment]
@timunie 

> I only have windows

I have windows too, and my repro (which includes powershell script) is for windows too. Noone confirm or rejected that the repro (https://github.com/AvaloniaUI/Avalonia/issues/5294#issuecomment-1790239889) has failed or works fine. This is sad... I don't know how to help this further. Noone guide me...

**Comment by buzzware** [comment]
I'm on an Intel Mac with Ventura.
This happens regularly on 11.0.5.
My solution used mixed latest nightly and 11.0.5.
If I built the solution, it would fail on a project on 11.0.5.
I just made all projects use 11.1.999-cibuild0041387-beta as @lsoft describes above, and the problem went away.

**Comment by timunie** [comment]
acc. to feedback this issue seems to be resolved in nightly. So I'm closing it. If someone has this happen on (upcomming) 11.1 or nightly, please ping me and I'll re-open the issue.

**Comment by lsoft** [comment]
@timunie sorry to bother, but I don't follow.... this issue had been closed and then reopened again because nightly builds fix the problem only with `-m:1` build key, which means sequentially building.

I had reported this in https://github.com/AvaloniaUI/Avalonia/issues/5294#issuecomment-1790239889

did I lost something important? parallel build is fixed now too? I can confirm we are experiencing troubles without `-m:1` key (in parallel building).

**Comment by TomEdwardsEnscape** [comment]
@lsoft can you reproduce the issue with a Nuget package built from #13840?

**Comment by lsoft** [comment]
> @lsoft can you reproduce the issue with a Nuget package built from #13840?

@TomEdwardsEnscape I'm happy to verify, but I can't find the appropariate nuget `42674` (https://github.com/AvaloniaUI/Avalonia/pull/13840#issuecomment-1841738563). The closest version is `42678` (https://nuget-feed-nightly.avaloniaui.net/packages/avalonia/11.1.999-cibuild0042678-beta). Could you clarify: does 42678 contains your fix?

**Comment by timunie** [comment]
@lsoft see feed url for All feed which includes PR: https://github.com/AvaloniaUI/Avalonia/pull/13840#issuecomment-1841738563 

Nightly doesnt

**Comment by lsoft** [comment]
@TomEdwardsEnscape @timunie looks like I can confirm that the fix is working! 👍 THANKS A LOT!!!

```
PS C:\temp> .\msbuild_mk.ps1
Iteration: 0, Now: 12/06/2023 15:14:17
Iteration: 1, Now: 12/06/2023 15:21:36
Iteration: 2, Now: 12/06/2023 15:22:22
Iteration: 3, Now: 12/06/2023 15:23:09
Iteration: 4, Now: 12/06/2023 15:23:43
Iteration: 5, Now: 12/06/2023 15:24:19
Iteration: 6, Now: 12/06/2023 15:24:53
Iteration: 7, Now: 12/06/2023 15:25:29
Iteration: 8, Now: 12/06/2023 15:26:05
Iteration: 9, Now: 12/06/2023 15:26:44
Iteration: 10, Now: 12/06/2023 15:27:18
Iteration: 11, Now: 12/06/2023 15:27:52
Iteration: 12, Now: 12/06/2023 15:28:26
Iteration: 13, Now: 12/06/2023 15:28:56
Iteration: 14, Now: 12/06/2023 15:29:29
Iteration: 15, Now: 12/06/2023 15:30:00
Iteration: 16, Now: 12/06/2023 15:30:36
Iteration: 17, Now: 12/06/2023 15:31:08
Iteration: 18, Now: 12/06/2023 15:31:43
Iteration: 19, Now: 12/06/2023 15:32:20
Iteration: 20, Now: 12/06/2023 15:32:52
Iteration: 21, Now: 12/06/2023 15:33:24
Iteration: 22, Now: 12/06/2023 15:33:56
Iteration: 23, Now: 12/06/2023 15:34:28
Iteration: 24, Now: 12/06/2023 15:35:00
Iteration: 25, Now: 12/06/2023 15:35:32
Iteration: 26, Now: 12/06/2023 15:36:04
Iteration: 27, Now: 12/06/2023 15:36:37
Iteration: 28, Now: 12/06/2023 15:37:09
```

[master_A2Z {3}.zip](https://github.com/AvaloniaUI/Avalonia/files/13580046/master_A2Z.3.zip)


but I noticed a strange message that I did not notice before:

![image](https://github.com/AvaloniaUI/Avalonia/assets/5988558/7f4acfdf-2718-46b4-863e-28916759eab3)

not sure if this have value.

Looking forward to see the fix in the next release, we are still compiling with `-m:1` and suffering from degrage compilation performance! 👍

**Comment by TomEdwardsEnscape** [comment]
Can you create a repro case which doesn't require the Android SDK?

**Comment by lsoft** [comment]
@TomEdwardsEnscape my naive repro (with Android deps!) is not working anymore, I ran it for 200 iterations with no build error. Now I want not only check msbuild process exit code, but to validate an output log for `CodeAnalysis` `CompileAvaloniaXamlTask` `The process cannot access the file` `error` and rerun.

Give me few days please

**Comment by lsoft** [comment]
@TomEdwardsEnscape I can confirm there is something wrong. I had an automated repro and I catched 3 errors like 

```
MSBUILD : Avalonia error AVLN9999: The process cannot access the file 'C:\Temp\master_A2Z\src\A2Z\BBBBBB\obj\Release\net7.0\BBBBBB.dll' because it is being used by another process. [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018: The "CompileAvaloniaXamlTask" task failed unexpectedly. [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018: System.IO.IOException: The process cannot access the file 'C:\Temp\master_A2Z\src\A2Z\BBBBBB\obj\Release\net7.0\refint\Avalonia\BBBBBB.dll' because it is being used by another process. [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018:    at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite) [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018:    at Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() in /_/src/Avalonia.Build.Tasks/CompileAvaloniaXamlTask.cs:line 46 [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
C:\Users\IAm\.nuget\packages\avalonia\11.1.999-cibuild0042674-beta\buildTransitive\AvaloniaBuildTasks.targets(150,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\Temp\master_A2Z\src\A2Z\BBBBBB\BBBBBB.csproj]
```
 I noticed 3 errors for 100 reruns, so I have 3% probability of failure, at least on my laptop. 

[compilation logs.zip](https://github.com/AvaloniaUI/Avalonia/files/13658297/error.compilation.zip)

automated repro (`master_a2z` can be taken from my previous posts):

```C#
        static void Main(string[] args)
        {
            Directory.SetCurrentDirectory(@"C:\temp\master_A2Z");

            File.WriteAllText(
                @"clean.bat",
                @$"dotnet clean > _.txt"
                );

            var i = 0;
            while (true)
            {
                Console.WriteLine("Iteration " + i);

                //-clp:"ErrorsOnly;Summary"
                ///bl:avalonia.binlog
                File.WriteAllText(
                    @"run.bat",
                    $"""
dotnet build --no-incremental -m /nr:false  /p:Platform="Any CPU";Configuration="Release";Optimize="true";DebugSymbols="true";DebugType="full"  >> {i}.txt
"""
                    );

                var rClean = Process.Start(
                    @"cmd.exe",
                    @"/C clean.bat"
                    );
                rClean.WaitForExit();

                var rRun = Process.Start(
                    @"cmd.exe",
                    @"/C run.bat"
                    );
                rRun.WaitForExit();

                var body = File.ReadAllText(i + ".txt").ToLower();
                if(body.Contains("codean") || body.Contains("xception"))
                {
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.WriteLine("^^^ suspicious!");
                    Console.ResetColor();
                }

                i++;
            }
        }
```

I have NO binlog yet, and I'm going to switch to binlog. working with binlog is diffucult a bit, I will need to replay it to get txt log, and then analyze that txt log.

I'll get back with new results.

**Comment by lsoft** [comment]
@TomEdwardsEnscape here is failed [avalonia2.binlog.zip](https://github.com/AvaloniaUI/Avalonia/files/13658730/avalonia2.binlog.zip)

I guess I may stop here. Let me know if you need additional assisting please.

**Comment by TomEdwardsEnscape** [comment]
The binlog immediately reveals the problem: `CompileAvaloniaXaml` is running repeatedly for `BBBBBB.csproj`. Each time for the same reason:

```
Input file "obj\Release\net7.0\refint\BBBBBB.dll" is newer than output file "obj\Release\net7.0\Avalonia\BBBBBB.dll".
```

I'll come back later and dig in more to work out why this is happening.

**Comment by TomEdwardsEnscape** [comment]
I think I fixed it, please try again with a new PR build. If the compiler has no work to do, it doesn't write any files. The task then copies the input files to the expected output paths by itself. But unlike MSBuild's `Copy` task, the `File.Copy` method doesn't alter the last modification time of the files it copies. So now the task does this, too.

**Comment by drweb86** [comment]
Hi guys, my compilation is broken because of the same bug. Do you plan to release fix anytime soon?

**Comment by timunie** [comment]
@drweb86 feel free to test https://github.com/AvaloniaUI/Avalonia/pull/13840#issuecomment-1854611182 and give us a feedback if that PR would fix your issue.

**Comment by drweb86** [comment]
I'll take a look in the evening.

**Comment by lsoft** [comment]
@TomEdwardsEnscape after 200 iterations I see no errors. Congratulations!

@timunie could you share in which version the fix will be posted?

**Comment by timunie** [comment]
@lsoft I'd think 11.1 should have it if nothing critical is against it.

**Comment by drweb86** [comment]
@timunie , bug returns.
"The process cannot access the file ....pdb' because it is being used by another process.

**Comment by drweb86** [comment]
Can it be another bug - i don't know

**Comment by TomEdwardsEnscape** [comment]
Which PDB? Which process cannot access it? What are you building? How are you building it?

Please build with Binary Logging enabled (how this is done depends on how you are building) and upload a .binlog file which shows the problem.

**Comment by TomEdwardsEnscape** [comment]
You aren't building with the PR branch. In fact you are quite far out of date:

```
AvaloniaVersion = 11.0.2
```

**Comment by drweb86** [comment]
> You aren't building with the PR branch. In fact you are quite far out of date:
> 
> ```
> AvaloniaVersion = 11.0.2
> ```
I have 11.0.6. See screenshot below.
![image](https://github.com/AvaloniaUI/Avalonia/assets/4712173/4c767861-480b-472f-951a-3003ae04828e)

I've tried to update to PR version from different  package repository, but since it did not fix the issue, i did not pushed it

**Comment by lsoft** [comment]
@drweb86 from my experience here, the only valuable help we can provide to the team is to attach binlog (which contains the error) taken from PR nuget (now `11.1.999-cibuild0042879-beta`). could you share such binlog please?

we are anxiously waiting your data to see if the error is fixed or not! :)

**Comment by TomEdwardsEnscape** [comment]
I see that you do have Avalonia 11.0.6, [but are defining `AvaloniaVersion` as 11.0.2 in this file](https://github.com/drweb86/butil/blob/fe293b44a3b3817078223e5677a2ce4e2ff9bfb3/sources/Directory.Build.props#L4).

Nevertheless, the binlog that you posted is from a build which does not contain the PR changes.

**Comment by timunie** [comment]
probably the Avalonia-All nuget-feed was missed or the prerelease check was not set. Then nuget falls back to nearest match iirc.

**Comment by drweb86** [comment]
> Nevertheless, the binlog that you posted is from a build which does not contain the PR changes.

One moment, I'll do it once again.

**Comment by drweb86** [comment]
> I see that you do have Avalonia 11.0.6
I updated it, please see branch 
https://github.com/drweb86/butil/tree/bug/%2311.1.999-cibuild0042879-beta

Also I see in 11.1.999-cibuild0042879-beta in logs, however I was thinking that there will be like Avalonia Build Services or something for building.

[Uploading binlog-v2.zip…]()

**Comment by TomEdwardsEnscape** [comment]
No Avalonia targets executed in that build, and the file which could not be written to (`BUtil.Core.deps.json`) is not accessed by any Avalonia target.

Your problems seem not to be related to Avalonia. It could be that another third-party target needs to be fixed up in a similar way that my PR fixes Avalonia.

**Comment by drweb86** [comment]
@TomEdwardsEnscape , thanks. I actually currently recloned repository to new folder and problem is fixed for ~20 runs. I don't understand why however, probably dlls are cached somehow.

**Comment by lsoft** [comment]
Guys, what's the status here? are we free to merge @TomEdwardsEnscape 's PR?

**Comment by TomEdwardsEnscape** [comment]
It's waiting for review from @kekekeks.
